### DISCUSSION TITLE
valhalla_ingest_transit, how long does it take?

### DISCUSSION THREAD

**Discussion by khamaileon** [valhalla/valhalla#4146] [Q&A] [answered]
Hi guys,

I'm in the process of adding public transport to my itineraries. To do this, I'm loading a [large GTFS file](https://navitia.opendatasoft.com/explore/dataset/fr-idf/files/d9722e11c1796205f33e8a02ae35ee43/download/) . But the valhalla_ingest_transit command has been running for nearly 20 hours and I have no idea when it's going to finish. I have 8 cores and 16GB of RAM. Do you have any idea how long it will take?

If this command can't be optimized in the future, perhaps it would be interesting to add a loading bar.

**Comment by nilsnolde** [comment]
What's the log saying?

**Reply by khamaileon** [reply]
That it's in progress.

Here are the latest warnings:

```
=> => # 2023/06/05 12:44:33.686330 [ERROR] Feed fr-idf, trip IDIDFM:TN:SNCF:d7e4b2ad-4e4e-4233-899c-106d307079ea can't be found or has no calendar.txt entry, skipping...   
 => => # 2023/06/05 12:56:38.960729 [INFO] Writing /opt/valhalla/transit/3/000/799/928.pbf.2                                                                                 
 => => # 2023/06/05 12:56:39.039839 [INFO] /opt/valhalla/transit/3/000/799/928.pbf.2 had 227446 stop pairs                                                                   
 => => # 2023/06/05 13:04:54.935712 [ERROR] Generated platform for station IDFM:StopPlaceEntrance:50171292                                                                   
 => => # 2023/06/05 14:37:44.478783 [INFO] Writing /opt/valhalla/transit/3/000/799/929.pbf                                                                                   
 => => # 2023/06/05 14:37:44.536347 [INFO] /opt/valhalla/transit/3/000/799/929.pbf had 181758 stop pairs
```

**Reply by nilsnolde** [reply]
hm, looks like a nasty feed, this line shouldn't ever happen:
`[ERROR] Generated platform for station IDFM:StopPlaceEntrance:50171292`

It's really far from being optimized in terms of ingestion performance. The PBF writing is sooo slow.. No one stepped up yet to fund optimizing it, so priority is kinda low atm unfortunately.

**Reply by khamaileon** [reply]
Thanks Nils. I don't need it to be perfect. If only one station or 2 doesn't work for some reason that's okay. Do you have any idea how long it usually takes on other GTFS?

**Reply by nilsnolde** [reply, marked as answer]
I didn't have any time yet to try around.. @gknisely ran a NYC feed, I believe it took 5 hours for the first stage. The second stage (valhalla_convert_transit) is MUCH faster.

**Reply by khamaileon** [reply]
Ok so that's normal, I think the Paris network is larger. Thank you!

**Comment by khamaileon** [comment]
I noticed that the first part of the script ran on a single core. Then all cores (and RAM 😬) were used. Then it went back down to 1 core.

**Reply by nilsnolde** [reply]
jep, that's right:) you can dive into the executable if you're interested

**Reply by khamaileon** [reply]
Yes I started looking but my C is so rusty that it would be faster for me to learn Rust instead 😆

**Reply by nilsnolde** [reply]
So, it's first parsing all the raw GTFS data (single thread, not so much time), then it's assembling the traffic PBFs (multi-thread, very long process), finally stitching the neighboring tiles (multithread, medium long process IIRC). So the more threads you throw at it, the better. We use a locked thread queue in those executables, so once it's down to one thread, it should be kinda done (although a single tile can be big, i.e. 2.5 km x 2-ish km, the tile resolution, can cover a lot of transit stuff in Paris I'd imagine). 

Would be great if you could report back how long it takes to build Paris, surely a stress test :)

**Reply by nilsnolde** [reply]
If you have 8 cores, you probably have 16 threads? Did you try with 16? With 16 GB RAM that should almost be fine, but no guarantees. At least it's hardly reading any tiles into the cache, and we dump the biggest stuff regularly to disk.

**Reply by khamaileon** [reply]
I have 8 CPUs and 2 threads per core. I haven't set the value of "mjolnir.concurrency" so I guess I'm at the max, i.e. 16 threads.

What's surprising is that since this morning only one of my CPUs is at 100%, whereas last night they were all at 100%. And I'm still in the "ingest_transit" threaded function because I'm still getting "can't be found or has no calendar.txt entry" errors.

**Reply by nilsnolde** [reply]
Off wow.. Then you're not even in the stitching phase yet.. Must be the last tile then. Can you `ll transit_pbf_dir/* -h` (whatever is in `mjonir.transit_dir`)? It'd be good to see what the tile sizes are (so far).

**Reply by khamaileon** [reply]
627M but doesn't seem to be changing at the moment.

**Reply by khamaileon** [reply]
Would it be worth changing the size of the tiles? If so, would it be compatible with the rest of the road network?

**Reply by nilsnolde** [reply]
> Would it be worth changing the size of the tiles?

Hm, yeah I guess it could, but probably minuscule considering the overall time (EDIT: hm, or maybe quite significant if you'd go pretty small, at least before the stitching, so there's somewhere a sweet spot). I'd try to change all tile size similarly.. I'm pretty sure it's not relevant that they're a factor of 4 apart, but of course only a common (integer) denominator of 180 & 360 is allowed (is denominator the right term? :sweat_smile: I guess you know what I mean).

**Reply by khamaileon** [reply]
Okay, I'll leave the tiles for now then. Good news: it finished in 25 hours. Bad news: I ran it in a Docker build and I don't know why I can't add any more commands.

#0 0.424 runc run failed: unable to start container process: exec: "/bin/sh": stat /bin/sh: no such file or directory

I'll restart outside of Docker. See you in 24 hours 🙄

**Reply by nilsnolde** [reply]
hard to support without knowing the chain of commands, config etc. either way, you shouldn't have have to restart the entire thing. you should've mapped the volume where the transit PBF are on your host I assume, then you can just use that for a native `valhalla_convert_transit` before starting the graph build.

**Reply by khamaileon** [reply]
Yes, I should have mapped a volume. It's all in a "corrupt" image now. But that's okay, I'll take this opportunity to test an another GTFS. I'll keep you posted. Thanks Nils.

**Comment by wildwestrom** [comment]
I'd like to add to this discussion. I'm using South Korean public transit data (modified from what you can obtain from [KTDB](https://www.ktdb.go.kr/www/index.do)). I must've started this thing about 3 hours ago and it's still going. Granted, it's for the entire country, but it feels terrible seeing only one cpu thread going and no progress indication.

Also, where's this log file supposed to be?

**Reply by nilsnolde** [reply]
Feel free to PR some progress logging, happy to review.

**Reply by kevinkreiser** [reply]
there is a PR to speed this up its in review at the moment #5134 feel free to give that branch a try
